js/test/Exchange/test.sharedMethods.js

```diff
@@ -0,0 +1,206 @@
+'use strict'
+
+const assert = require ('assert');
+const Precise = require ('../../base/Precise');
+
+function logTemplate (exchange, method, entry) {
+    return ' <<< ' + exchange.id + ' ' + method + ' ::: ' + exchange.json (entry) + ' >>> ';
+}
+
+function reviseStructureKeys (exchange, method, entry, format, emptyNotAllowedFor = []) {
+    // define common log text
+    const logText = logTemplate (exchange, method, entry);
+    // ensure item is not null/undefined/unset
+    assert (entry, 'item is null/undefined' + logText);
+    // get all expected & predefined keys for this specific item and ensure thos ekeys exist in parsed structure
+    if (Array.isArray (format)) {
+        assert (Array.isArray (entry), 'entry is not an array' + logText);
+        const realLength = entry.length;
+        const expectedLength = format.length;
+        assert (realLength === expectedLength, 'entry length is not equal to expected length of ' + expectedLength.toString () + logText);
+        for (let i = 0; i < format.length; i++) {
+            const is_in_array = exchange.inArray (i, emptyNotAllowedFor);
+            if (is_in_array) {
+                assert ((entry[i] !== undefined), i.toString () + ' index is undefined, but is is was expected to be set' + logText);
+            }
+            // because of other langs, this is needed for arrays
+            try {
+                assert (areSameTypes (exchange, entry, i, format), i.toString () + ' index does not have an expected type ' + logText);
+            } catch (ex) {
+                assert (false, i.toString () + ' index is missing from structure' + logText);
+            }
+        }
+    } else {
+        assert (typeof entry === 'object', 'entry is not an object' + logText);
+        const keys = Object.keys (format);
+        for (let i = 0; i < keys.length; i++) {
+            const key = keys[i];
+            const keyStr = key.toString ();
+            assert ((key in entry), keyStr + ' key is missing from structure' + logText);
+            if (exchange.inArray (key, emptyNotAllowedFor)) {
+                // if it was in needed keys, then it should have value.
+                assert (entry[key] !== undefined, key + ' key has an null value, but is expected to have a value' + logText);
+            }
+            assert (areSameTypes (exchange, entry, key, format), key + ' key is neither undefined, neither of expected type' + logText);
+        }
+    }
+}
+
+function areSameTypes (exchange, entry, key, format) {
```

Comment:
also I'm not yet sure if they should receive the `exchange` because we're creating another dependency that seems unnecessary

Reply:
can you elaborate, what dependency/drawback does it have?

Reply:
Idk but let's say we decide to completely refactor `exchange` to something different why would some generic helper methods break with that? They should not be coupled with the exchange


Second let's say our product evolves and we add something other than the exchange, (a terminal, graphs, whatever) If we want to to reuse these methods there we can't because they require the exchange

Reply:
if `exchange` changes, most of lib and even `test.xyz` files will need to be rewritten. and what a problem would be just to change few methods? even what you say, might not happen within near years, and even if it happens, this will be least problem imo. passing exchange gives more flexibility to generate uniform logs or even use `exchange.safeValue` and etc..
actually, i don't see it as a problem atm.
but as you decide.